casper/src/main/scala/coop/rchain/casper/Estimator.scala

```diff
@@ -21,103 +19,103 @@ object Estimator {
   implicit val decreasingOrder = Ordering[Long].reverse
 
   def tips[F[_]: Monad: BlockStore](
-      blockDag: BlockDag,
+      blockDag: BlockDagRepresentation[F],
       lastFinalizedBlockHash: BlockHash
   ): F[IndexedSeq[BlockMessage]] =
-    Estimator.tips[F](blockDag, lastFinalizedBlockHash, blockDag.latestMessages.map {
-      case (validator, block) => (validator, block.blockHash)
-    })
+    for {
+      latestMessageHashes <- blockDag.latestMessageHashes
+      result              <- Estimator.tips[F](blockDag, lastFinalizedBlockHash, latestMessageHashes)
+    } yield result
 
   /**
     * When the BlockDag has an empty latestMessages, tips will return IndexedSeq(genesis)
     */
   def tips[F[_]: Monad: BlockStore](
-      blockDag: BlockDag,
+      blockDag: BlockDagRepresentation[F],
       lastFinalizedBlockHash: BlockHash,
       latestMessagesHashes: Map[Validator, BlockHash]
   ): F[IndexedSeq[BlockMessage]] = {
-    @tailrec
     def sortChildren(
-        blocks: IndexedSeq[BlockHash],
-        childMap: Map[BlockHash, Set[BlockHash]],
+        blocks: List[BlockHash],
```

Comment:
I don't like that this has become a `List`. The complexity for sorting a `List` is much higher than sorting an `IndexedSeq` because you don't have constant-time random element access. Since the whole point of this function is to do sorting I think `IndexedSeq` will be significantly more performant, which is why it used `IndexedSeq` originally.

Reply:
The reason why I decided to change it to `List` is heavy usage of `traverse`, `flatTraverse` in `Estimator` in this PR which can't be easily invoked on `IndexedSeq`. I have looked through the code and failed to understand why `IndexedSeq` is needed here instead of a `List`. @KentShikama also did not know the reason  when I contacted him about this. I think you are wrong about sorting complexity here: both `List` and `IndexedSeq` use default `SeqLike.sorted` implementation which just collects the collection into an array, sorts it by using `java.util.Arrays.sort` and builds the collection back.